lms/djangoapps/courseware/views/views.py

```diff
@@ -749,40 +743,118 @@ def _progress(request, course_key, student_id):
+        and certs_api.cert_generation_enabled(course_key)
+    )
+
+    show_cert_web_view = False
+    download_url = None
+    cert_web_view_url = None
+
+    if enrollment_mode == CourseMode.AUDIT:
+        return CertData(
+            CertificateStatuses.audit_passing,
+            'Your enrollment: Audit track',
+            'You are enrolled in the audit track for this course. The audit track does not include a certificate.',
+            download_url,
+            show_cert_web_view,
+            cert_web_view_url
+        )
+
+    if not show_generate_cert_btn:
+        return None
+
+    if certs_api.is_certificate_invalid(student, course_key):
+        return CertData(
+            CertificateStatuses.invalidated,
+            'Your certificate has been invalidated',
+            'Please contact your course team if you have any questions.',
+            download_url,
+            show_cert_web_view,
+            cert_web_view_url
+        )
+
+    cert_downloadable_status = certs_api.certificate_downloadable_status(student, course_key)
+    # If the learner is in verified modes and the student did not have
+    # their ID verified, we need to show message to ask learner to verify their ID first
+    missing_required_verification = enrollment_mode in CourseMode.VERIFIED_MODES and \
+        not SoftwareSecurePhotoVerification.user_is_verified(student)
+
+    if cert_downloadable_status['is_downloadable']:
+        cert_status = CertificateStatuses.downloadable
+        title = 'Your certificate is available'
+        msg = 'You can keep working for a higher grade, or request your certificate now.'
+        download_url = cert_downloadable_status['download_url']
         if certs_api.has_html_certificates_enabled(course_key, course):
             if certs_api.get_active_web_certificate(course) is not None:
-                context.update({
-                    'show_cert_web_view': True,
-                    'cert_web_view_url': certs_api.get_certificate_url(course_id=course_key, uuid=cert_status['uuid']),
-                })
+                show_cert_web_view = True
+                cert_web_view_url = certs_api.get_certificate_url(
+                    course_id=course_key, uuid=cert_downloadable_status['uuid']
+                )
             else:
-                context.update({
-                    'is_downloadable': False,
-                    'is_generating': True,
-                    'download_url': None
-                })
+                download_url = None
+                cert_status = CertificateStatuses.generating
+                title = "We're working on it..."
+                msg = "We're creating your certificate. You can keep working in your courses and a link " \
+                      "to it will appear here and on your Dashboard when it is ready."
+
+        return CertData(cert_status, title, msg, download_url, show_cert_web_view, cert_web_view_url)
```

Comment:
I'd push this `return` statement up into the body of the `if` and `else` blocks.